daemons/ipa-otpd/passkey.c

```diff
@@ -0,0 +1,825 @@
+{
+    (void)vctx; /* Unused */
+    struct child_ctx *child_ctx;
+
+    child_ctx = verto_get_private(ev);
+
+    free(child_ctx);
+}
+
+static int set_fd_nonblocking(int fd)
+{
+    int flags;
+    int ret;
+
+    flags = fcntl(fd, F_GETFL, 0);
+    if (flags == -1) {
+        ret = errno;
+        return ret;
+    }
+
+    if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) == -1) {
+        ret = errno;
+        return ret;
+    }
+
+    return 0;
+}
+
+#ifndef PASSKEY_CHILD_PATH
+#define PASSKEY_CHILD_PATH "/usr/libexec/sssd/passkey_child"
+#endif
+
+static int do_passkey_response(struct otpd_queue_item *item)
+{
+    int ret;
+    pid_t child_pid;
+    int pipefd_to_child[2] = { -1, -1};
+    int pipefd_from_child[2] = { -1, -1};
+    /* Up to 50 arguments to the helper supported. The amount of arguments
+     * is controlled inside this function. Right now max used is below 20 */
+    char *args[50] = {NULL};
+    size_t args_idx = 0;
+    struct child_ctx *child_ctx;
+    char *pk = NULL;
+
+    child_ctx = calloc(sizeof(struct child_ctx), 1);
+    if (child_ctx == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+    child_ctx->item = item;
+
+    pk = ipa_passkey_get_public_key(item->user.ipaPassKey,
+                           item->passkey->data_in->data.response.credential_id);
+    if (pk == NULL) {
+        ret = EINVAL;
+        otpd_log_err(ret, "No matching public key found for [%s]",
+                     item->passkey->data_in->data.response.credential_id);
+        goto done;
+    }
+
+    args[args_idx++] = PASSKEY_CHILD_PATH;
+    args[args_idx++] = "--verify-assert";
+    args[args_idx++] = "--domain";
+    args[args_idx++] = item->passkey->domain;
+    args[args_idx++] = "--key-handle";
+    args[args_idx++] = item->passkey->data_in->data.response.credential_id;
+    args[args_idx++] = "--public-key";
+    args[args_idx++] = pk;
+    args[args_idx++] = "--cryptographic-challenge";
+    args[args_idx++] = item->passkey->data_in->data.response.cryptographic_challenge;
+    args[args_idx++] = "--auth-data";
+    args[args_idx++] = item->passkey->data_in->data.response.authenticator_data;
+    args[args_idx++] = "--signature";
+    args[args_idx++] = item->passkey->data_in->data.response.assertion_signature;
+
+    ret = pipe(pipefd_from_child);
+    if (ret == -1) {
+        ret = errno;
+        goto done;
+    }
+    ret = pipe(pipefd_to_child);
+    if (ret == -1) {
+        ret = errno;
+        goto done;
+    }
+
+    child_pid = fork();
+
+    if (child_pid == 0) { /* child */
+        close(pipefd_to_child[1]);
+        ret = dup2(pipefd_to_child[0], STDIN_FILENO);
+        if (ret == -1) {
+            exit(EXIT_FAILURE);
+        }
+
+        close(pipefd_from_child[0]);
+        ret = dup2(pipefd_from_child[1], STDOUT_FILENO);
+        if (ret == -1) {
+            exit(EXIT_FAILURE);
+        }
+
+        execv(args[0], args);
+        exit(EXIT_FAILURE);
```

Comment:
shouldn't this be `_exit()` instead of `exit()`? I think the same applies to older code as well? See example in `pipe(2)` man page.

Reply:
Actually, it matters for `vfork()`, for `fork()` we can use `exit()`.